### PR TITLE
Fix `opam exec` on Windows

### PR DESCRIPTION
**Pull request by AltGr**
split out from #4504

Expose `OpamProcess.create_process_env`, which has the correct way of calling 
cygwin executables from a Windows native opam (without unwanted expansion of 
arguments, etc.)

And use that for `opam exec` (instead of `Unix.execv`) so that it has the 
correct behaviour.

### CODE REVIEW

File: src/core/opamProcess.mli [hunk 1]
```diff
@@ -214,3 +214,10 @@ type 'a job = 'a Job.Op.job
 (**/**)
 val set_resolve_command :
   (?env:string array -> ?dir:string -> string -> string option) -> unit
+
+(** Like Unix.create_process_env, but with correct escaping of arguments when
+    invoking cygwin executables *)
```
Comment (dra27) [inline src/core/opamProcess.mli:219]:
```suggestion
    invoking cygwin executables from a native Windows executable. *)
```

File: src/core/opamProcess.ml [hunk 2]
```diff
@@ -310,6 +310,13 @@ let set_resolve_command =
     resolve_command_fn := resolve_command
 let resolve_command cmd = !resolve_command_fn cmd
 
+let create_process_env cmd args env cin cout cerr =
+  let resolve cmd = OpamStd.Option.default cmd @@ resolve_command cmd in
+  if Sys.win32 && OpamStd.Sys.is_cygwin_variant (resolve cmd) = `Cygwin then
+    cygwin_create_process_env cmd args env cin cout cerr
+  else
+    Unix.create_process_env cmd args env cin cout cerr
```
Comment (dra27) [inline src/core/opamProcess.ml:318]: FWIW:
```suggestion
let create_process_env =
  if Sys.win32 then
    fun cmd ->
      let resolved_cmd = resolve_command cmd in
      if OpamStd.(Option.map_default Sys.is_cygwin_variant `Native resolved_cmd) = `Cygwin then
        cygwin_create_process_env cmd
      else
        Unix.create_process_env
  else
    Unix.create_process_env
```
The `Sys.win32` hoist is a personal meta-programming foible of mine (i.e. everything based on `Sys.win32` should be done once, as the value can't change!) but getting partial application `OpamProcess.create_process_env "foo"` to do the resolve immediately is useful/better.

**Review by dra27 — COMMENTED** [review]
Couple of suggestions...

File: src/core/opamProcess.mli (hunk 1 above)
Comment (dra27) [reply src/core/opamProcess.mli:219]: (and with apologies for changing my mind on the wording _again_!)

File: master_changes.md [hunk 3]
```diff
@@ -7,7 +7,7 @@ New option/command/subcommand are prefixed with ◈.
   *
 
 ## Global CLI
-  *
+  * Fix `opam exec` on native Windows when calling cygwin executables (@AltGr)
```
Comment (rjbou) [inline master_changes.md:10]:
```suggestion
  * Fix `opam exec` on native Windows when calling cygwin executables [#4588 @AltGr]
```

**Comment by rjbou** [comment]
Should be fixed now

**Pull request merged**